src/main/kotlin/uk/gov/justice/digital/hmpps/sentenceplan/services/GoalService.kt

```diff
@@ -190,55 +190,87 @@ class GoalService(
-      type = when {
-        updatedGoal.status == GoalStatus.FUTURE && this.goal!!.status == GoalStatus.REMOVED -> GoalNoteType.READDED
-        updatedGoal.status == GoalStatus.ACTIVE && this.goal!!.status == GoalStatus.REMOVED -> GoalNoteType.READDED
-        updatedGoal.status == GoalStatus.REMOVED -> GoalNoteType.REMOVED
-        updatedGoal.status == GoalStatus.ACHIEVED -> GoalNoteType.ACHIEVED
-        else -> GoalNoteType.PROGRESS
-      }
+    goalEntity.notes.add(
+      GoalNoteEntity(note = updatedGoal.note!!, goal = goalEntity, type = GoalNoteType.ACHIEVED),
+    )
+    goalEntity.status = updatedGoal.status!!
+    goalEntity.statusDate = LocalDateTime.now()
+
+    return goalRepository.save(goalEntity)
+  }
+
+  @Transactional
+  fun removeGoal(goalUuid: UUID, updatedGoal: Goal): GoalEntity {
+    if (updatedGoal.status != GoalStatus.REMOVED) {
+      throw ValidationException("Updated Goal status must be REMOVED but was ${updatedGoal.status}")
+    } else if (updatedGoal.note.isNullOrEmpty()) { // note is mandatory for a removed goal
+      throw ValidationException("Updated goal note must not be empty")
     }
-    goalEntity.notes.add(goalNoteEntity)
-
-    // If the goal was re-added then we need to set the order of the goal to the highest value
-    // so that it appears last in the plan overview.
-    if (goalNoteEntity.type == GoalNoteType.READDED) {
-      val planVersionEntity: PlanVersionEntity
-      try {
-        planVersionEntity = planVersionRepository.findByUuid(goalEntity.planVersion!!.uuid)
-      } catch (e: EmptyResultDataAccessException) {
-        throw NotFoundException("A Plan with this UUID was not found: $goalEntity.planVersion!!.uuid")
-      }
 
-      val highestGoalOrder = planVersionEntity.goals.maxByOrNull { g -> g.goalOrder }?.goalOrder ?: 0
-      goalEntity.goalOrder = highestGoalOrder + 1
-
-      // also need to set the new targetDate
-      if (updatedGoal.targetDate != null) {
-        goalEntity.targetDate = LocalDate.parse(updatedGoal.targetDate)
-        goalEntity.status = GoalStatus.ACTIVE
-        goalEntity.statusDate = LocalDateTime.now()
-      } else {
-        goalEntity.targetDate = null
-        goalEntity.status = GoalStatus.FUTURE
-        goalEntity.statusDate = LocalDateTime.now()
-      }
+    val goalEntity = goalRepository.getGoalByUuid(goalUuid)
+
+    goalEntity.notes.add(
+      GoalNoteEntity(note = updatedGoal.note, goal = goalEntity, type = GoalNoteType.REMOVED),
+    )
+    goalEntity.status = updatedGoal.status!!
+    goalEntity.statusDate = LocalDateTime.now()
+
+    return goalRepository.save(goalEntity)
+  }
+
+  @Transactional
+  fun reAddGoal(goalUuid: UUID, updatedGoal: Goal): GoalEntity {
+    if (updatedGoal.status != GoalStatus.ACTIVE && updatedGoal.status != GoalStatus.FUTURE) {
+      throw ValidationException("Updated Goal status must be ACTIVE or FUTURE but was ${updatedGoal.status}")
+    }
+
```

Comment:
Is a note also mandatory for re-adding a goal? If so we should put a check in similar to the above, using `updatedGoal.note.isNullOrEmpty()`

Reply:
Double checked, yes it is mandatory. Resolved in [1e4b4d7](https://github.com/ministryofjustice/hmpps-sentence-plan/pull/199/commits/1e4b4d76850be0d3b55bc82a5a91be5f74093f2c)